src/system/SystemObject.h

```diff
@@ -229,10 +253,27 @@ inline T * ObjectPool<T, N>::TryCreate(Layer & aLayer)
 {
     T * lReturn = nullptr;
     unsigned int lIndex;
-#if CHIP_SYSTEM_CONFIG_PROVIDE_STATISTICS
-    unsigned int lNumInUse = 0;
-#endif
 
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+    for (lIndex = 0; lIndex < mObjects.size(); ++lIndex)
+    {
+        if (mObjects[lIndex]->TryCreate(aLayer, sizeof(T)))
+        {
+            lReturn = mObjects[lIndex].get();
+            break;
+        }
+    }
+
+    if (lReturn == nullptr && mObjects.size() < N)
+    {
+        mObjects.push_back(std::unique_ptr<T>(new T()));
+
+        if (mObjects.back()->TryCreate(aLayer, sizeof(T)))
```

Comment:
In which circumstances can this fail? Should we `push_back` in such a case?

Reply:
TryCreate will fail if the object has already been used(this->mSystemLayer != nullptr).  Currently, the actual number of object in use and actual number of object in the pool are not same, the object allocated previously could be re-used after it has been released.

We always try to find if there is object available to use before we push_back an new object to the pool. This is to keep syntax aligned between heap implementation and static allocation implementation.

Reply:
We should free the objects when their lifetime ends, not leave them for reuse. This will ensure sanitizers work out of the box.

Reply:
I did consider this design, but the object is released by its user directly instead of via ObjectPool currently, to achieve what you described, we need to introduce an additional reference to the owner ObjectPool within Object to class , and free this object via the wrapping unique pointer. And again, this is only apply to the heap based implementation. Any better idea?

Reply:
> I did consider this design, but the object is released by its user directly instead of via ObjectPool currently

How does that work if the pool is holding a `unique_ptr` to the object?   Are we basically ending up in a situation where once you have allocated M objects there's no way to get the pool to be sized less than M short of clearing the entire pool?

Perhaps the answer is that we _should_ in fact change the allocation/freeing patterns for these objects, and maybe start with the various pools we have that are not SystemObject if the SystemObject case is particularly hard?

Reply:
You are right, as long as the object is added to the pool, it can only be reused (after released), but not removed from the pool until the program end. The pool is released and all objects are freed by their unique_ptr. The benefit is we actually only allocate the maximum number of objects we actually use and keep the interface unchanged with static allocation.

Otherwise we need to change the allocation/freeing API for both heap and bss based implementations, which allocate and release object both through pool.

Reply:
> The benefit is we actually only allocate the maximum number of objects we actually use

Why is this a benefit?  The maximum number might be much larger than the average number....

> Otherwise we need to change the allocation/freeing API for both heap and bss based implementations, 

Yes.

Reply:
I have updated the code to free the objects when their lifetime ends.

Reply:
The change to call Release on the sPool objects is pragmatic and simple.  I still think a linked list could be simpler and more performant (no more iterating over the whole vector).  But this looks like it will work.